src/Orleans/Runtime/OutsideRuntimeClient.cs

```diff
@@ -109,21 +96,45 @@ public IStreamProviderRuntime CurrentStreamProviderRuntime
             Justification = "MessageCenter is IDisposable but cannot call Dispose yet as it lives past the end of this method call.")]
         public OutsideRuntimeClient(ClientConfiguration cfg, bool secondary = false)
         {
-            this.typeCache = new TypeMetadataCache();
-            this.assemblyProcessor = new AssemblyProcessor(this.typeCache);
-            this.grainFactory = new GrainFactory(this, this.typeCache);
-
             if (cfg == null)
             {
                 Console.WriteLine("An attempt to create an OutsideRuntimeClient with null ClientConfiguration object.");
                 throw new ArgumentException("OutsideRuntimeClient was attempted to be created with null ClientConfiguration object.", "cfg");
             }
 
+            var services = new ServiceCollection();
+            services.AddSingleton(cfg);
+            services.AddSingleton<IServiceProvider>(sp => sp);
+            services.AddSingleton<TypeMetadataCache>();
+            services.AddSingleton<AssemblyProcessor>();
+            services.AddSingleton<IRuntimeClient>(this);
+            services.AddSingleton<GrainFactory>();
+            services.AddFromExisting<IGrainFactory, GrainFactory>();
+            services.AddFromExisting<IInternalGrainFactory, GrainFactory>();
+            services.AddSingleton<ClientProviderRuntime>();
+            services.AddFromExisting<IMessagingConfiguration, ClientConfiguration>();
+            services.AddSingleton<IGatewayListProvider>(
+                sp => ActivatorUtilities.CreateInstance<GatewayProviderFactory>(sp).CreateGatewayListProvider());
+            services.AddSingleton<SerializationManager>();
+            services.AddSingleton<ClientStatisticsManager>();
+            services.AddSingleton<MessageFactory>();
+            services.AddSingleton<RequestContextUtil>();
+            services.AddSingleton<StreamProviderManager>();
+            services.AddFromExisting<IStreamProviderManager, StreamProviderManager>();
+            services.AddSingleton<CodeGeneratorManager>();
+
+            this.ServiceProvider = cfg.ServiceConfigurator(services);
+
+            this.InternalGrainFactory = this.ServiceProvider.GetRequiredService<IInternalGrainFactory>();
+            this.ClientStatistics = this.ServiceProvider.GetRequiredService<ClientStatisticsManager>();
+            this.SerializationManager = this.ServiceProvider.GetRequiredService<SerializationManager>();
+            this.messageFactory = this.ServiceProvider.GetService<MessageFactory>();
```

Comment:
Understood.  Huge PR, so I'm quite ok with pushing this to a later refactor.  Just wanted to make sure I understood why it was this way.  Having said that, the more of this that you can structure well in this pass, the more well formed the code will be if we don't get back to this immediately.  I'm not confident we've a good record of doing cleanup passes yet.